src/python/pants/backend/python/rules/pex.py

```diff
@@ -75,6 +79,14 @@ class CreatePex:
   input_files_digest: Optional[Digest] = None
 
 
+@dataclass(frozen=True)
+class CreatePexFromTargetClosure:
+  """Represents a request to create a PEX from the closure of a set of targets."""
+  build_file_addresses: BuildFileAddresses
```

Comment:
Why plural? It feels like I'd expect the structure of "stuff that should go in a single pex" to be fully represented in BUILD files and modelled as dependencies? What's your use-case here?

Reply:
This merely captures what the "create binary" rule was already doing. Refactored out because I need the same functionality for "create AWS lambda". So that's two use-cases.

Generally, it seems to me that "turn this set of targets (and by implication, their dependencies) into a pex" seems like a sensible piece of functionality.

Or did you mean why not just a *single* address? True, I don't currently have a use-case for multiple targets. But we could, for example, use this to bring in pytest when running tests. That is, instead of having the pytest version living in an option it lives in a BUILD file (and addressed by an option). That's off the top of my head, admittedly

Reply:
Yeah, asking about the multiple addresses thing - I can't think of any non-contrived use-cases...

Reply:
Agreed that this should probably be a singular `BFA`. 

Note that the original wrapped the single `BFA` into the plural form in the `await Get[TransitiveHydratedTargets]`. I recommend changing this dataclass to just use a single BFA, then in line 160 do:

```python
transitive_hydrated_targets = await Get[TransitiveHydratedTargets](BuildFileAddresses, (request.build_file_address,))

Reply:
Why though? There appears to be no downside?

Reply:
I don't see why we would restrict the functionality, when there is at least one plausible use-case for it.

Reply:
> I don't see why we would restrict the functionality

I try to err on the side of keeping functionality limited to the problem we have today while allowing for a future design to generalize it more. 

Agreed that there's a possible future scenario where we want multiple addresses - if that ends up being the case, we can simply refactor this dataclass to allow plural BFAs.

For now, we only ever use a single BFA and it's easier to reason about the semantics of operating on only one target rather than figuring out the semantics of multiple targets being turned into one Pex (at least, it's easier for to reason about). It's less complex now but still easily changed in the future to be more generalized if we need it to be.

Reply:
I'm going to push back on this a bit. I don't think this adds any complexity, and it opens the door for new use-cases, possibly including the one I mentioned. To "err on the side of keeping functionality limited to the problem we have today" doesn't take into account third party plugins and functionality they may need.

Reply:
Note: I don't care enough about this issue to argue beyond basically this comment, so don't really mind if this gets ignored, but wanted to explain my thinking :)

My concern is that it may lead to implicit dependencies which aren't reflected in the build graph. Right now, I rely on BUILD files being correct in order to generate a deployable binary. And I can trust that when I run other tasks; if I write a task which checks for unused dependencies, or transitive deps that should be direct deps, or looks up the licences of my 3rdparty deps, or whatever, I can trust that Pants can statically analyse the transitive closure of deps (either because they're in a BUILD file, or because they're synthetically injected, or potentially because of dependency inference), and know that I have complete information to operate from.

The only point I can see of being able to specify multiple BFAs is to say "I know I don't say in a BUILD file that I don't depend on a logging implementation (or whatever), but this target will only work if you happen to specify a logging implementation as a root alongside it". And that takes away my ability to statically reason about a dependency graph - it adds extra context I need to know outside of my source code in order to correctly build this deployable.

We see this in the JVM world at the moment - people get really surprised when the resolve of their 3rdpary deps is different depending on what other targets they specified as roots. This API design is guiding people towards making that same mistake.

Ideally, imo, there would be one way of specifying dependencies (and maybe super-specific hooks for overrides, like the pytest rule has a flag for overriding the requirements there), and this feels like a _really_ general hammer which takes away a pretty simple one-way-to-do-something guarantee.

Reply:
Well, this isn't something that's exposed to users, only to rule authors. And a pex isn't necessarily a deployable, we use it in a variety of cases as a behind-the-scenes construct. 

We already have an example where we provide extra dependencies, namely pytest. We don't require tests to explicitly depend on pytest (unless they import from pytest), we inject it as a runtime dep because we happen to use pytest as a test runner. Now admittedly we don't currently do this via a build target, but via options, but the principle holds (and we could imagine replacing our option-based solution with build targets a la BUILD.tools).

Another example might codegen, where there's a runtime dep that we inject, so that people don't have to have their thrift_library targets explicitly depend on language-specific 3rdparty libs.